admin/services/check-start.service.js

```diff
@@ -122,4 +195,54 @@ checkStartService.pupilLogin = async function (pupilId) {
   return { checkCode: check.checkCode, questions }
 }
 
+/**
+ * Query the DB and put the info into messages suitable for placing on the prepare-check queue
+ * The message needs to contain everything the pupil needs to login and take the check
+ * @param checkFormAllocationIds
+ * @return {Promise<Array>}
+ */
+checkStartService.prepareCheckQueueMessages = async function (checkFormAllocationIds) {
+  if (!checkFormAllocationIds) {
+    throw new Error('checkFormAllocationIds is not defined')
+  }
+
+  if (!Array.isArray(checkFormAllocationIds)) {
+    throw new Error('checkFormAllocationIds must be an array')
+  }
+
+  const messages = []
+  const checkFormAllocations = await checkFormAllocationDataService.sqlFindByIdsHydrated(checkFormAllocationIds)
+  const sasExpiryDate = moment().add(config.Tokens.sasTimeOutHours, 'hours')
+  const sasToken = sasTokenService.generateSasToken(sasTokenService.queueNames.CHECK_COMPLETE, sasExpiryDate)
+
+  for (let o of checkFormAllocations) {
+    const config = await configService.getConfig({id: o.pupil_id}) // ToDo: performance note: this does 2 sql lookups per pupil. Optimise!
```

Comment:
It seems like there could be some optimisations made by requesting all/batches of pupil data at once instead of synchronously waiting for each pupil's config to load inside a loop. Also, is there a limit to the number of messages that can be input into Azure queue at any time, because this could potentially be rejected if there is a limit?

I note there is also a "todo" here - I feel that ideally we should not be merging in todos? Could this not be dealt with now?

Reply:
this is in a separate card along with some other optimisations